### ISSUE TITLE
customize dataset class and data loader 

### ISSUE THREAD

**Issue by fatemeyousefii** [ultralytics/ultralytics#15175]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

hi , i have a dataset that structed as bellow , this is my dataset.yaml :
train: /home/training/Yousefi/new/Physical_Cells/train
val: /home/training/Yousefi/new/Physical_Cells/val

nc: 3  # Number of classes
names: ['1', '2', '3']  # List of class names :/home/training/Yousefi/physical_cells/
  ├── train/
  │   ├── images/
  │   │   ├── image1.png
  │   │   ├── image2.png
  │   ├──labels/
  │       ├── image1.txt
  │       ├── image2.txt
  ├── val/
      ├── images/
      │   ├── image1.png
      │   ├── image2.png
      ├──labels/
          ├── image1.txt
          ├── image2.txt
and each labeled is saved in .txt file ...now i want to write a customize dataset class ( not using default) ...here is my code : import os
import torch
from torch.utils.data import Dataset, DataLoader
from PIL import Image
from torchvision import transforms

import os
import torch
from torch.utils.data import Dataset
from PIL import Image
from torchvision import transforms

class CustomYOLODataset(Dataset):
    def __init__(self, img_dir, label_dir, img_size=640, transform=None):
        self.img_dir = img_dir
        self.label_dir = label_dir
        self.img_size = img_size
        self.transform = transform

        self.img_files = [f for f in os.listdir(img_dir) if f.endswith('.png')]
        self.label_files = [f.replace('.png', '.txt') for f in self.img_files]

    def __len__(self):
        return len(self.img_files)

    def __getitem__(self, idx):
        img_path = os.path.join(self.img_dir, self.img_files[idx])
        label_path = os.path.join(self.label_dir, self.label_files[idx])

        image = Image.open(img_path).convert('RGB')

        boxes, labels = [], []
        if os.path.exists(label_path):
            with open(label_path, 'r') as f:
                for line in f:
                    class_id, x_center, y_center, width, height = map(float, line.split())
                    labels.append(int(class_id))

                    boxes.append([x_center, y_center, width, height])

        if self.transform:
            image = self.transform(image)

        # Handle case with no boxes
        if len(boxes) == 0:
            boxes = torch.empty((0, 4), dtype=torch.float32)
            labels = torch.empty((0,), dtype=torch.long)
        else:
            boxes = torch.tensor(boxes, dtype=torch.float32)
            labels = torch.tensor(labels, dtype=torch.long)

        return image, {'boxes': boxes, 'labels': labels}

def get_transform(train):
    transforms_list = []
    if train:
        transforms_list.append(transforms.RandomHorizontalFlip(0.5))
    transforms_list.append(transforms.ToTensor())
    return transforms.Compose(transforms_list)

def collate_fn(batch):
    images, targets = zip(*batch)
    images = torch.stack(images, dim=0)
    targets = {
        'boxes': [t['boxes'] for t in targets],
        'labels': [t['labels'] for t in targets]
    }
    return images, targets

def create_data_loader(img_dir, label_dir, batch_size, img_size, train=True):
    dataset = CustomYOLODataset(img_dir, label_dir, img_size, transform=get_transform(train))
    data_loader = DataLoader(dataset, batch_size=batch_size, shuffle=train, num_workers=4, collate_fn=collate_fn)
    return data_loader

import torch
from ultralytics import YOLO

def train_yolo_v5():
    # Set up the model and device
    model = YOLO('yolov5s.pt')

    # Training parameters
    training_args = {
        'task': 'detect',
        'mode': 'train',
        'model': 'yolov5s.pt',
        'data': 'dataset.yaml',  # Use your custom dataset YAML file here
        'epochs': 5,
        'batch': 4,
        'imgsz': 640,
        'device': 'cuda:0',
        'optimizer': 'auto',
        'verbose': True,
        'patience': 100,
        'lr0': 0.01,
        'name': 'train',
        'amp': True,
        'rect': False,
        'cos_lr': False,
        'overlap_mask': True,
        'mask_ratio': 4,
        'augment': False,
        'fliplr': 0.5,
        'auto_augment': 'randaugment',
        'hsv_h': 0.015,
        'hsv_s': 0.7,
        'hsv_v': 0.4,
        'translate': 0.1,
        'scale': 0.5,
        'mosaic': 1.0,
        'mixup': 0.0,
        'save': True,
        'project': 'runs/detect/train'
    }

    # Run training
    model.train(**training_args)

# Example usage
train_yolo_v5()
.... i want to train my dataset based on train_step but when i run this code i got this error : WARNING ⚠️ no labels found in detect set, can not compute metrics without labels .... i think the out put of get_item() in my dataset class is not compatible with yolo ...how should the output would be? and how can i solve this problem ? please notice that i want to use custom dataset class and loader with train_step .thanks

**Comment by pderrenger** [comment]
@fatemeyousefii it looks like you're on the right track with your custom dataset class and data loader. The error message "no labels found in detect set, cannot compute metrics without labels" suggests that the YOLO model is not receiving the labels correctly from your custom dataset.

YOLO expects the dataset to return images and corresponding labels in a specific format. For object detection, the labels should include bounding boxes and class IDs. Your `__getitem__` method seems to handle this, but there might be a mismatch in how the data is being passed to the model.

Ensure that the labels are correctly formatted and that the bounding boxes are in the expected format (i.e., normalized coordinates). Additionally, verify that the `collate_fn` function correctly batches the data.

If the issue persists, you might want to test your custom dataset class independently to ensure it returns the expected output. You can also compare the output format with the default YOLO dataset class to identify any discrepancies.

For further assistance, please refer to the [YOLO Common Issues Guide](https://docs.ultralytics.com/guides/yolo-common-issues/) for more troubleshooting tips.

**Comment by chuang091** [comment]
Hi, @fatemeyousefii Do you tried to delete the cache file in the inages or labels? If that doesn't work can you show a part of your `label.txt` ?
